chore(build): replace Grunt with Rollup

epreston opened the pull request:
Replaces the build system with Rollup.  This will allow the code to be transpiled with Babel for pure es5 or es6. Provides a base for the rewrite.

Fixes: https://github.com/playcanvas/engine/issues/5399
Unblocks: https://github.com/playcanvas/engine/issues/5521

Other Bits:
- updates spine 3.8 to version with patches. <-- might resolve some issues
- removes favicon warning on examples to reduce stress while testing

Notes:

Vendor runtime repo is located here: https://github.com/EsotericSoftware/spine-runtimes

From 4.0, the latest patch version is on NPM.  for 3.6 and 3.8, its available as branches in the repo.
To automate updates, npm versions are alaised to make multiple versions of the same package available to this project.  See package.json (spine40, spine41).

They provide iife builds and instructions to modify them for use as ESM exports.  Not ideal for an engine integration or plugin (its more targeted at inclusion in html headers).  For the rewrite, its better to generate esm from the typescript source for each version with:

```
esbuild --bundle spine-ts/core/index.ts --target=es6 --sourcemap --outfile=spine.x.x.js --format=es
```

It is possible to add Babel now, but, the configuration will be replaced very shortly.

Next steps:

Reorganise the file structure of examples so we can swap to "serve" as dev server.  

Make examples attractive like engine's examples using pcui (but keeping it simple).  Might be worth making a tiny "examples" template that can be used across projects.

The next steps for the code are generating ESM bundles for 3.6 and 3.8 from the typescript source because vendor only publishes processed typescript to npm from 4.0. At that point, we need to remove the "context" (it needs to be top level undefined), introduce babbel to support language correct es5 and es6 targets (if translation is required), and flip the module setting accordingly.

I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

epreston committed:
Rollup Build

use prebundle semver for space savings
Add export wrapper for 4.1
ensure a context is supplied to iife so "this" is not overwritten to top level undefined.

epreston committed:
Update spine 3.8 to version with patches

epreston committed:
Modify builds as noted by vendor in readme

Vendor provides instructions for using this in esm builds.

epreston committed:
Update prebuild modules

epreston committed:
Remove favicon warning

epreston committed:
Remove Grunt Build System

mvaligursky commented:
That sounds great! We'll check it out next week, as we have a long weekend already.

willeastcott commented on src/spine.js:
```diff
@@ -1,3 +1,7 @@
+// ./wrapper.js will be aliased to other versions during builds
+import { spine } from './wrapper.js';
+import * as semver from './semver/index.js';
```
Can we simply import `semver` from `node_modules` now?

epreston replied:
That is.  That selection wrapper imports from semver, saves 20 KB on the build (see top of semver readme), and is needed for rollup/node_resolve to do its thing.

But that brings up another point, I can drop the commands for pre-generated semver iife and the files in contrib.

epreston replied:
completed in: 292b754

willeastcott commented on package.json:
```diff
@@ -33,18 +33,20 @@
   },
   "scripts": {
-    "build": "grunt",
+    "build": "rollup --config",
```
Minor point but all our other `build` scripts use the `rollup -c` syntax.

epreston replied:
completed in: 9c7531d

willeastcott commented on rollup.config.mjs:
```diff
@@ -0,0 +1,71 @@
+        output: [
+            {
+                file: `${outputDir + name}.js`,
+                format: 'cjs',
```
Shouldn't this be `'umd'`? For instance, see: https://github.com/playcanvas/engine/blob/128b0dfbe0d3cf7a75a0463169b6eccf0e3286a4/rollup.config.mjs#L339

epreston replied:
It does not have value.. yet.  That's for the rewrite.  There's no module exports yet.  The implementation is still based on the ordered execution of iife functions.

- cjs will just push the processed files, you get the nearly same bundled output with treeshake
- umd will put a wrapper function with no name, no exports, that is stripped away in the minified builds.  In short, it indents text and gets removed.

It has a place in the next steps:

The next steps are generating ESM bundles for 3.6 and 3.8 from the typescript source because vendor only publishes processed typescript to npm from 4.0.  At that point, we need to remove the "context" (it needs to be top level undefined), introduce babbel to support language correct es5 and es6 targets (if translation is required), and flip this module setting accordingly.

To get the same drag and drop workflow in the editor. The final product might continue to be an iife function that relies on engine being registered as a global called "pc".  I'm looking closer at engine plugins while I wait.

willeastcott replied:
OK, sounds good.

epreston committed:
Remove prebuilt semver iife

iife module is no longer required

epreston committed:
Update package.json

one less word to slowly sound out

epreston committed:
Make 4.0 Support a 3 Line Chore

Add alias for 4.0.  versions below this were not published to npm.  They are available as branches in the spine-core repo.

epreston committed:
Add 4.0 build stub

willeastcott commented on rollup.config.mjs:
```diff
@@ -0,0 +1,75 @@
+        name: 'playcanvas-spine.3.8',
+        lib: 'contrib/spine-ts/build/3.8/spine-core.js'
+    },
+    // {
+    //     name: 'playcanvas-spine.4.0',
+    //     lib: 'spine40' // may require wrapper or adjustment
+    // },
```
What's this? Are you intending on adding 4.0 support as well or something?

epreston replied:
It's 2 lines of code and some testing.  I am working up the courage.  

Want to make sure the build system is ready in this PR.

willeastcott replied:
OK, I'll hold off on merging for now then. Let me know when you think you're done. 😄

epreston replied:
Im done, not doing 4.0 in this PR.  Ready to merge.  

Just wanted to make the build config clearer by showing the version progression.

willeastcott replied:
Oh OK. I'd rather not merge commented out or incomplete stuff to main though. Can you delete this for now and if you want to add in a separate PR, do it there?

epreston replied:
ok, just a note to others, removing now.  Keen to start on examples PR.

epreston replied:
completed in: 1a543df

epreston committed:
Update rollup.config.mjs

move example to example

willeastcott commented on package.json:
```diff
@@ -33,20 +33,21 @@
-    "semver": "^7.5.4"
+    "rollup": "^3.28.1",
+    "semver": "^7.5.4",
+    "spine40": "npm:@esotericsoftware/spine-core@4.0",
```
OK, last change request...just remove this line and I'll merge.

epreston replied:
The alias ?  Ok.  Guess it makes the 4.0 PR clearer.

willeastcott replied:
Yeah, exactly. It'll keep that PR more 'atomic'. 👍

epreston replied:
Thank you. Completed in: 014a17c

epreston committed:
Remove spine40

"spine40": "npm:@esotericsoftware/spine-core@4.0"

Will be a part of the 4.0 PR.

Make sure we are at latest PlayCanvas code standard.

epreston committed:
Fix lint warning PlayCanvas Eslint 1.7.0

willeastcott commented on package.json:
```diff
@@ -33,20 +33,20 @@
+    "@rollup/plugin-commonjs": "^25.0.4",
+    "@rollup/plugin-node-resolve": "^15.2.1",
+    "@rollup/plugin-terser": "^0.4.3",
+    "esbuild": "^0.19.2",
```
Oh, actually, is `esbuild` needed any more?

epreston replied:
That is part of the build system for consuming vendor typescript.

willeastcott replied:
Can you point me at where it's invoked? I don't see it in the repo....

epreston replied:
Future bits, I updated the notes up top.  Might be worth pulling in the official "@rollup/plugin-typescript" now as well so we don't need to touch the build system in other PRs ?

willeastcott replied:
If it's needed for future bits, I would include it in a future PR. And at that point, we can debate whether it's required. But if the current state of the repo doesn't require `esbuild`, I would remove it. Again, this is me having a strong preference to keep PRs minimal, complete and atomic.

epreston replied:
Your call.  Build changes in one build PR or when its used.  We know we'll be consuming typescript next, that's what the vendors code really is.

willeastcott replied:
I definitely prefer 'when it's used'. 😄

epreston replied:
Then it goes.  Completed in 40db1ce

epreston committed:
Remove esbuild

Second half of build system will be in another PR.

willeastcott commented:
Yay. LGTM.
